[docs] GraphQL client documentation and configuration fixes (#714)
ExpediaGroup/graphql-kotlin: docs/client/client-overview.md

---
id: client-overview
title: Client Overview
---

`graphql-kotlin-client` is a lightweight type-safe GraphQL HTTP client. Type-safe data models are generated at build time
by the GraphQL Kotlin [Gradle](https://expediagroup.github.io/graphql-kotlin/docs/plugins/gradle-plugin) and
[Maven](https://expediagroup.github.io/graphql-kotlin/docs/plugins/maven-plugin) plugins.

`GraphQLClient` is a thin wrapper on top of [Ktor HTTP Client](https://ktor.io/clients/index.html) and supports fully

> asynchronous non-blocking communication. Client is highly customizable and can be configured with any supported HTTP

Reviewer: ```suggestion
asynchronous non-blocking communication. It is highly customizable and can be configured with any supported Ktor HTTP
```

                <groupId>com.expediagroup</groupId>
                <artifactId>graphql-kotlin-maven-plugin</artifactId>
                <version>${graphql-kotlin.version}</version>
                <executions>
                    <execution>
                        <id>introspect-schema</id>
                        <goals>
                            <goal>introspectSchema</goal>
                        </goals>
                        <configuration>
                            <endpoint>http://localhost:8080/graphql</endpoint>
                        </configuration>
                    </execution>
                    <execution>
                        <id>generate-client</id>
                        <goals>
                            <goal>generateClient</goal>
                        </goals>
                        <configuration>
                            <packageName>com.example.generated</packageName>
                            <schemaFile>${project.build.directory}/schema.graphql</schemaFile>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
        </plugins>
    </build>
</project>
```

> See [graphql-kotlin-client-example](https://github.com/dariuszkuc/graphql-kotlin-client-example) project for complete

Author: I could move this project under `/examples` but they won't be run during the build, i.e. in order to build clients they need a running server.

working examples of Gradle and Maven based projects.

## Generating GraphQL Client

> By default, GraphQL Kotlin build plugins will attempt to generate GraphQL clients from all query files located under

Reviewer: ```suggestion
By default, GraphQL Kotlin build plugins will attempt to generate GraphQL clients from all `*graphql` files located under
```

> `src/main/resources`. Queries are validated against target GraphQL schema which can be manually provided, retrieved by

Reviewer: ```suggestion
`src/main/resources`. Queries are validated against the target GraphQL schema, which can be manually provided, retrieved by
```

the plugins through introspection (as configured in examples above) or downloaded directly from a custom SDL endpoint.
See our documentation for more details on supported [Gradle tasks](https://expediagroup.github.io/graphql-kotlin/docs/plugins/gradle-plugin#tasks)
and [Maven Mojos](https://expediagroup.github.io/graphql-kotlin/docs/plugins/maven-plugin#goals).

When creating your GraphQL queries make sure to always specify an operation name and name the files accordingly. Each

> one of your query files will result in a generation of a corresponding Kotlin file with a class matching your operation

Reviewer: ```suggestion
one of your query files will generate a corresponding Kotlin file with a class matching your operation
```

name that will act as a wrapper for all corresponding data classes. For example, given `HelloWorldQuery.graphql` with
`HelloWorldQuery` as the operation name, GraphQL Kotlin plugins will generate a corresponding `HelloWorldQuery.kt` file
with a `HelloWorldQuery` class under the configured package.

For example, given a simple schema

```graphql
type Query {
  helloWorld: String
}
```

And a corresponding `HelloWorldQuery.graphql` query

```graphql
query HelloWorldQuery {
  helloWorld
}
```

Plugins will generate following client code

```kotlin
package com.example.generated

import com.expediagroup.graphql.client.GraphQLClient
import com.expediagroup.graphql.client.GraphQLResult
import kotlin.String

const val HELLO_WORLD_QUERY: String = "query HelloWorldQuery {\n    helloWorld\n}"

class HelloWorldQuery(
  private val graphQLClient: GraphQLClient
) {
  suspend fun execute(): GraphQLResult<HelloWorldQuery.HelloWorldQueryResult> =
      graphQLClient.execute(HELLO_WORLD_QUERY, "HelloWorldQuery", null)

>   data class HelloWorldQueryResult(

Reviewer: To simplify the naming should we also just generate a data class called `Result` or `Response`? Since it is already nested I think it is pretty clear the type and that means we can do less string concatenation too.

`HelloWorldQuery.Result()`

Author: sure

    val helloWorld: String
  )
}
```

> Generated class requires an instance of `GraphQLClient` and exposes a single `execute` suspendable method that executes

Reviewer: ```suggestion
Generated classes requires an instance of `GraphQLClient` and exposes a single `execute` suspendable method that executes
```

> the underlying GraphQL operation using provided client.

Reviewer: ```suggestion
the underlying GraphQL operation using the provided client.
```
